modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsManagerImpl.java

```diff
@@ -89,86 +157,468 @@ public IgniteStatisticsRepository statisticsRepository() {
+            Map<UUID, List<Integer>> reqNodes
+    ) {
+        Map<UUID, StatsCollectionAddrRequest> res = new HashMap<>(reqNodes.size());
+        for (Map.Entry<UUID, List<Integer>> reqNode : reqNodes.entrySet()) {
+            UUID reqId = UUID.randomUUID();
+            StatsCollectionRequest colReq = new StatsCollectionRequest(colId, reqId, Collections.singletonMap(keyMsg,
+                    reqNode.getValue().stream().mapToInt(Integer::intValue).toArray()));
+            StatsCollectionAddrRequest addrReq = new StatsCollectionAddrRequest(colReq, reqNode.getKey());
+            res.put(reqId, addrReq);
+        }
+        return res;
+    }
+
+    /**
+     * TODO
+     */
+    public void stop() {
+        if (statMgmtPool != null) {
+            List<Runnable> unfinishedTasks = statMgmtPool.shutdownNow();
+            if (!unfinishedTasks.isEmpty())
+                log.warning(String.format("%d statistics collection request cancelled.", unfinishedTasks.size()));
+        }
     }
 
     /**
      * Collect partition level statistics.
      *
      * @param tbl Table to collect statistics by.
-     * @param selectedColumns Columns to collect statistics by.
+     * @param partIds Array of partition ids to collect statistics by.
+     * @param selectedCols Columns to collect statistics by.
+     * @param cancelled Supplier to check if collection was cancelled.
      * @return Collection of partition level statistics by local primary partitions.
      * @throws IgniteCheckedException in case of error.
      */
-    private Collection<ObjectPartitionStatisticsImpl> collectPartitionStatistics(GridH2Table tbl, Column[] selectedColumns)
-            throws IgniteCheckedException {
+    private Collection<ObjectPartitionStatisticsImpl> collectPartitionStatistics(
+            GridH2Table tbl,
+            int[] partIds,
+            Column[] selectedCols,
+            Supplier<Boolean> cancelled
+    ) throws IgniteCheckedException {
         List<ObjectPartitionStatisticsImpl> tblPartStats = new ArrayList<>();
         GridH2RowDescriptor desc = tbl.rowDescriptor();
         String tblName = tbl.getName();
+        GridDhtPartitionTopology topology = tbl.cacheContext().topology();
+        AffinityTopologyVersion topologyVersion = topology.readyTopologyVersion();
+
+        for (int partId : partIds) {
+            GridDhtLocalPartition locPart = topology.localPartition(partId, topologyVersion, false);
+            if (locPart == null)
+                continue;
+
+            if (cancelled.get()) {
+                if (log.isInfoEnabled())
+                    log.info(String.format("Canceled collection of object %s.%s statistics.", tbl.identifier().schema(),
+                            tbl.identifier().table()));
+
+                return null;
+            }
 
-        for (GridDhtLocalPartition locPart : tbl.cacheContext().topology().localPartitions()) {
             final boolean reserved = locPart.reserve();
 
             try {
```

Comment:
I think MOVING partitions can be safely skipped here.
There is no need to wait for preloader synchronously, you can subscribe to the future and reschedule statistic gathering once it's done.

Reply:
skipped